llvm/lib/Target/AArch64/AArch64Combine.td

```diff
@@ -131,6 +131,13 @@ def ext: GICombineRule <
   (apply [{ applyEXT(*${root}, ${matchinfo}); }])
 >;
 
+def fullrev: GICombineRule <
+  (defs root:$root, shuffle_matchdata:$matchinfo),
+  (match (G_SHUFFLE_VECTOR $src, $src1, $src2, $mask):$root,
+         [{ return matchFullRev(*${root}, MRI); }]),
```

Comment:
Predicate seems like it can go in the generic combiner

Reply:
I think the types should be OK - v4i32 are either handled here or by perfect shuffles and the other types should be handled by other combines as far as I can tell.